File: Container/MXDocumentsList.m
//
//  MXDocumentsList.m
//  Lines
//
//  Created by Paolo Bosetti on 26/09/14.
//  Copyright (c) 2014 University of Trento. All rights reserved.
//

#import "MXDocumentsList.h"

@implementation MXDocumentsList

+ (MXDocumentsList *)documentsListAtPath:(NSString *)path
                                forTypes:(NSArray *)types {
  MXDocumentsList *docs = [[MXDocumentsList alloc] init];
  [docs scanFolder:path forTypes:types];
  return docs;
}

- (instancetype)init {
  self = [super init];
  if (self) {
    self.folderPath =
        [NSHomeDirectory() stringByAppendingPathComponent:@"Documents"];
    self.types = [NSMutableArray
        arrayWithObjects:MXLogDocument.class, MXMovieDocument.class, nil];
    self.list = [[NSMutableDictionary alloc] init];
  }
  return self;
}

- (void)scanFolder:(NSString *)path forTypes:(NSArray *)types {
  if (path != nil)
    self.folderPath = path;
  if (types != nil)
    self.types = (NSMutableArray *)types;
  NSFileManager *manager = [NSFileManager defaultManager];
  NSArray *allItems =
      [manager contentsOfDirectoryAtPath:self.folderPath error:nil];

  for (Class k in self.types) {
    NSMutableArray *docsList =
        [k filterFileNames:allItems atPath:self.folderPath];
    [self.list setObject:docsList forKey:[k humanName]];
  }
}

- (MXDocument *)documentAtIndexPath:(NSIndexPath *)indexPath {
  NSString *key = [_types[indexPath.section] humanName];
  NSInteger idx = indexPath.row;
  MXDocument *doc = self.list[key][idx];
  return doc;
}

- (void)removeDocumentAtIndexPath:(NSIndexPath *)indexPath {
  NSString *key = [_types[indexPath.section] humanName];
  NSInteger idx = indexPath.row;
  MXDocument *doc = self.list[key][idx];
  [_list[key] removeObjectAtIndex:idx];
  [[NSFileManager defaultManager] removeItemAtPath:doc.filePath error:nil];
}

- (NSString *)sectionAtIndex:(NSInteger)index {
  return [_list allKeys][index];
}

@end

File: Container/MXDocumentsList.swift
//
//  MXDocumentsList.swift
//  Lines
//
//  Created by Paolo Bosetti on 02/10/14.
//  Copyright (c) 2014 University of Trento. All rights reserved.
//

import Foundation

class MXDocumentsList: NSObject {
  var list: Dictionary<String,[MXDocument]> = [:]
  var types: Array<MXDocument.Type> = []
  var folderPath: String = ""
  
  class func documentsListAtPath(path: String, forTypes: [MXDocument.Type]) -> MXDocumentsList {
    var result: MXDocumentsList?
    result?.scanFolder(path, forTypes: forTypes)
    return result!
  }
    
  override init() {
    self.folderPath = "\(NSHomeDirectory())Documents"
    self.types = [MXLogDocument.self , MXMovieDocument.self]
  }
  
  func scanFolder(path: String?, forTypes:[MXDocument.Type]?) {
    if (path != nil) {
      self.folderPath = path!
    }
    if (forTypes != nil) {
      self.types = forTypes!
    }
    let manager = NSFileManager.defaultManager()
    let allItems = manager.contentsOfDirectoryAtPath(self.folderPath, error: nil) as [String]
    var docsList: [MXDocument] = []
    for k in self.types {
      let n = k.humanName()
      docsList = k.filterFileNames(allItems, atPath: self.folderPath)
      self.list[n] = docsList
    }
  }
  
  func documentAtIndexPath(indexPath: NSIndexPath) -> MXDocument {
    return MXDocument()
  }
  
  func removeDocumentAtIndexPath(indexPath: NSIndexPath) {
    let key = self.types[indexPath.section].humanName()
    let idx = indexPath.row
    let doc = self.list[key]?[idx];
    list[key]?.removeAtIndex(idx)
    NSFileManager.defaultManager().removeItemAtPath(doc!.filePath, error:nil);
  }
  
  func sectionAtIndex(index: Int) -> String {
    return self.list.keys.array[index]
  }
  
}